### ISSUE TITLE
[EPIC] Release Lyo 2.4.0 

### ISSUE THREAD

**Issue by jamsden** [eclipse/lyo#2]
Let's use this issue to discuss what should be included in the release and track its completion.

The primary elements for consideration in Release 2.4.0 may include:

### Upgrade to Sirius 5

This includes various improvements of Lyo Designer that Yash has been waiting on Sirius 5 to introduce. e.g., delete a resource in a specification diagram, popups a dialog requesting if you want to delete from diagram or delete from model.

This is complete and included in 2.4.0-SNAPSHOT

### ~~Enable embedding Lyo Core as an eclipse plugin~~

This involves creating a ‘lyo-core-bundle’, which will be an OSGi bundle built with Apache Felix Maven Bundle Plugin (BND) plus will have all of its dependencies vendored via the Embed-Transitive instruction of the BND plugin.

This is complete and included in 2.4.0-SNAPSHOT

### ~~Export Domain Vocabularies to RDF~~

Allows exporting Lyo Designer Domain specifications as RDF vocabularies and/or OSLC ResourceShapes or SHACL. Import is not completed yet. This does not include the the ability to model SHACL shapes. 

Ready for review and merge (was blocked waiting for lyo-core-bundle).

### Lyo Validation

Validates RDF resources against constraints defined either as SHACL or OSLC Resource Shapes.

This is currently waiting for approval of two CQs:

- [x] [CQ 16618](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=16618)
- [x] [CQ 13908](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=13908)

CQ 13908 is for the shaclex library with includes over 80+ dependencies. Either this will take a long time, or we could consider distributing Lyo Validation with Type A due diligence in order to avoid holding up the release.

eclipse/Lyo, like any eclipse (sub) project can only deliver one thing under one due diligence type. So we can't deliver OSLC4J and Validation in the same project under two different due diligence types.

There are a few options for resolving this:

1. ~~Remove Validation from release 2.4.0 (note that it was already, perhaps erroneously included in 2.3.0, which exposes OSLC4J 2.3.0 with Type A content - this could be a problem).~~ This artefact is considered released and will be kept.
2. ~~Hold up release 2.4.0 until CQ 13908 is approved for Type B~~ Eclipse IP team does not give us a clear timeline. **Also, we need to explicitly trigger this review!**
3. ✅ Decouple Validation from eclipse/Lyo, refactor the packages and artifact ids to remove references to Lyo, and distribute outside of eclipse. **@yashkhatri is assigned to https://github.com/eclipse/lyo.validation/issues/24**

My recommendation would be #1 unless there is some compelling business reason for including Validation in 2.4.0.

### TRS Consumer and Provider Library

This provides a framework for building TRS Providers on an adapted data source. The TRS consumer is a sample consumer that provides a SPARQL endpoint on data provided by one or more TRS providers. 

This is work in progress, some additional refactoring might need to be done. It was not released in 2.3.0, and has a dependency on Validation.

**Recommend this not be included in 2.4.0**, but is a candidate for 2.5.0. There is no reason to hold up the 2.4.0 release to complete this work.

### Open issues to be resolved

There are issues currently open in [eclipse/Bugzilla against 2.4.0](https://bugs.eclipse.org/bugs/buglist.cgi?bug_status=UNCONFIRMED&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&columnlist=bug_id%2Ccomponent%2Cshort_short_desc%2Cbug_status%2Cassigned_to_realname%2Creporter_realname%2Clongdescs.count%2Cchangeddate&list_id=17509407&product=Lyo&query_format=advanced&target_milestone=2.4.0).

New list of items to be included in 2.4.0:

- [x] :construction: [Code generate (optional) swagger documentation by jadelkhoury · Pull Request #40 · eclipse/lyo.designer](https://github.com/eclipse/lyo.designer/pull/40)
- [ ] :construction: ~~[Adding support for abstract types by neormx · Pull Request #29 · eclipse/lyo.core](https://github.com/eclipse/lyo.core/pull/29)~~
- [x] [Downgrade the version · Issue #24 · eclipse/lyo.validation](https://github.com/eclipse/lyo.validation/issues/24)

Old list of items to be include in 2.4.0:

 - [ ] :construction: [393073 – NPE in the hasNext method in class OslcQueryResult](https://bugs.eclipse.org/bugs/show_bug.cgi?id=393073)
- [ ] :construction: [528403 – Architecture Management Resource not compliant to OSLC AM 2.0](https://bugs.eclipse.org/bugs/show_bug.cgi?id=528403)
- [ ] :construction: [518808 – AllowedValues.getValues fails if there is only a single allowed value](https://bugs.eclipse.org/bugs/show_bug.cgi?id=518808)
- [ ] :construction: [Bundling the client as osgi bundle · Issue #6 · eclipse/lyo.client](https://github.com/eclipse/lyo.client/issues/6)
- [ ] [534210 – Consumer Secret OAuth passwords saved in base64](https://bugs.eclipse.org/bugs/show_bug.cgi?id=534210)
- [ ] [531428 – RmUtil return Resource Shape where XSD:int(http://www.w3.org/2001/XMLSchema#int) is null.](https://bugs.eclipse.org/bugs/show_bug.cgi?id=531428)
- [ ] [518809 – Lyo fails with NPE in the presence of badly-formed OSLC property definitions](https://bugs.eclipse.org/bugs/show_bug.cgi?id=518809)
- [x] [526202 – Allow Lyo libraries to be reused with Lyo OSGI](https://bugs.eclipse.org/bugs/show_bug.cgi?id=526202)
 - [x] [531949 – Support JSON-LD](https://bugs.eclipse.org/bugs/show_bug.cgi?id=531949)
- [x] [529618 – Override all OslcClient#getResource(...) in OslcOAuthClient](https://bugs.eclipse.org/bugs/show_bug.cgi?id=529618)
- [x] [Revert Jena Libs Dependency Change in Pom · Issue #34 · eclipse/lyo.store](https://github.com/eclipse/lyo.store/issues/34)

**Legend**

:construction: means a Pull Request has been submitted and the review/rework is pending.

### Release notes

[Draft on hackmd](https://hackmd.io/pXqQXqzLStSI4R7vu7PY2w?both)

**Comment by neormx** [comment]
Hope "Bug 531949 - Support JSON-LD" could be closed and included.

**Comment by berezovskyi** [comment]
yep @neormx and thanks for its initial implementation!

**Comment by berezovskyi** [comment]
Update: me and @jadelkhoury did a review today. Merged his patch and decided to postpone @neormx's https://github.com/eclipse/lyo.core/pull/29 till 2.5.0 unless we are really fast with it.

I have removed the "2.4.0 (A)" release entry and renamed "2.4.0 (B)" to simply "2.4.0". I also set the release date to Nov 7.

Eclipse seems down and once it's back I will file an IP log review.
